p2p/protocol/holepunch/coordination.go

```diff
@@ -0,0 +1,386 @@
+	if err := wr.WriteMsg(msg); err != nil {
+		return 0, nil, fmt.Errorf("failed to write CONNECT message to initator: %w", err)
+	}
+
+	// Read SYNC message
+	msg.Reset()
+	if err := rd.ReadMsg(msg); err != nil {
+		return 0, nil, fmt.Errorf("failed to read message from initator: %w", err)
+	}
+	if t := msg.GetType(); t != pb.HolePunch_SYNC {
+		return 0, nil, fmt.Errorf("expected SYNC message from initiator but got %d", t)
+	}
+	return time.Since(tstart), obsDial, nil
+}
+
+func (hs *Service) handleNewStream(s network.Stream) {
+	rp := s.Conn().RemotePeer()
+	rtt, addrs, err := hs.incomingHolePunch(s)
+	if err != nil {
+		hs.tracer.ProtocolError(rp, err)
+		log.Debugw("error handling holepunching stream from", rp, "error", err)
+		s.Reset()
+		return
+	}
+	s.Close()
+
+	// Hole punch now by forcing a connect
+	pi := peer.AddrInfo{
+		ID:    rp,
+		Addrs: addrs,
+	}
+	hs.tracer.StartHolePunch(rp, addrs, rtt)
+	log.Debugw("starting hole punch with", rp)
+	start := time.Now()
+	err = hs.holePunchConnect(pi)
+	dt := time.Since(start)
+	hs.tracer.EndHolePunch(rp, dt, err)
+	if err != nil {
+		log.Debugw("hole punching peer", rp, "failed after", dt, "error", err)
+	} else {
+		log.Debugw("hole punching peer", rp, "succeeded after", dt)
+	}
+}
+
+func (hs *Service) holePunchConnect(pi peer.AddrInfo) error {
+	holePunchCtx := network.WithSimultaneousConnect(hs.ctx, "hole-punching")
+	forceDirectConnCtx := network.WithForceDirectDial(holePunchCtx, "hole-punching")
+	dialCtx, cancel := context.WithTimeout(forceDirectConnCtx, dialTimeout)
+	defer cancel()
+
+	hs.tracer.HolePunchAttempt(pi.ID)
+	err := hs.host.Connect(dialCtx, pi)
+	if err == nil {
+		log.Debugw("hole punch with peer", pi.ID, "successful")
+	}
+	log.Debugw("hole punch attempt with peer", pi.ID, "failed:", err)
+	return err
+}
+
+func isRelayAddress(a ma.Multiaddr) bool {
+	_, err := a.ValueForProtocol(ma.P_CIRCUIT)
+	return err == nil
+}
+
+func addrsToBytes(as []ma.Multiaddr) [][]byte {
+	bzs := make([][]byte, 0, len(as))
+	for _, a := range as {
+		bzs = append(bzs, a.Bytes())
+	}
+	return bzs
+}
+
+func addrsFromBytes(bzs [][]byte) []ma.Multiaddr {
+	addrs := make([]ma.Multiaddr, 0, len(bzs))
+	for _, bz := range bzs {
+		a, err := ma.NewMultiaddrBytes(bz)
+		if err == nil {
+			addrs = append(addrs, a)
+		}
+	}
+	return addrs
+}
+
+type netNotifiee Service
+
+func (nn *netNotifiee) Connected(_ network.Network, v network.Conn) {
+	hs := (*Service)(nn)
+	dir := v.Stat().Direction
+
+	// Hole punch if it's an inbound proxy connection.
+	// If we already have a direct connection with the remote peer, this will be a no-op.
+	if dir == network.DirInbound && isRelayAddress(v.RemoteMultiaddr()) {
+		p := v.RemotePeer()
+		hs.activeMx.Lock()
+		hs.closeMx.RLock()
+		closed := hs.closed
+		_, active := hs.active[p]
+		if !active && !closed {
+			hs.refCount.Add(1)
+			hs.active[p] = struct{}{}
+		}
+		hs.closeMx.RUnlock()
+		hs.activeMx.Unlock()
+
+		if active || closed {
+			return
+		}
```

Comment:
Wrapping this all in a single `beginHolePunch` function may make it simpler.

Reply:
Oh yes, it does!